Lib/test/test_hashlib.py

```diff
@@ -1196,6 +1196,29 @@ def test_file_digest(self):
             with open(os_helper.TESTFN, "wb") as f:
                 hashlib.file_digest(f, "sha256")
 
+    @unittest.skipUnless(support.check_sanitizer(thread=True), "only meaningful on free-threading")
+    def test_gh_128657(self):
+        def test_hashlib_sha256():
+            hash_obj = hashlib.sha256()
+
+        def closure(barrier):
+            barrier.wait()
+            test_hashlib_sha256()
+
+        num_workers = 40
+        num_runs = 20
+
+        for i in range(num_runs):
+            barrier = threading.Barrier(num_workers)
+            thrds = []
+
+            for i in range(num_workers):
+                thrds.append(thrd := threading.Thread(target=closure, args=(barrier,)))
+                thrd.start()
+
+            for thrd in thrds:
+                thrd.join()
```

Comment:
We have a utility for starting threads that will simplify this, and we also need to make sure that the thread doesn't raise an exception (threads that fail will not propagate an error to the main thread).

```suggestion
        with threading_helper.catch_threading_exception() as cm:
            thrds = [threading.Thread(target=closure, args=(barrier,))]
            with threading_helper.start_threads(thrds):
                pass
             
             if cm.exc_value:
                 raise cm.exc_value
```